[DPE-9311] Initial user documentation (#24)
canonical/valkey-operator: docs/how-to/manage-passwords.md

# How to manage passwords

> In order to read or write data in Valkey, we need to authenticate ourselves.

Reviewer: Suggestion: Let's add an introduction and improve wording (using "in order to" is not recommended).

```suggestion
This guides provides instructions for creating, updating, and otherwise managing passwords.

To read or write data in Valkey, we need to authenticate ourselves.
```

Author: updated

> In order to read or write data in Valkey, we need to authenticate ourselves.

Reviewer: Suggestion: Let's add an introduction and improve wording (using "in order to" is not recommended).

```suggestion
This guides provides instructions for creating, updating, and otherwise managing passwords.

To read or write data in Valkey, we need to authenticate ourselves.
```

Author: done

```

Now you can use the password to access Valkey. Select the IP address for one of the units
and check the current health with this command:

```text
$ valkey-cli -h 10.1.44.126 -p 6379
10.1.44.126:6379> AUTH charmed-operator changeme
OK
10.1.44.126:6379> ping
PONG
```

## Update the password

To update your user-configured password, simply update the value of the secret. Here's an example:

```text
juju update-secret passwords charmed-operator=moresecurepassword
```

After running this command, Charmed Valkey will immediately update the password.
After the deployment has settled again, you can no longer use the old password to
access Valkey. Instead, you will receive an error similar to this:

```text
$ valkey-cli -h 10.1.44.126 -p 6379
10.1.44.126:6379> AUTH charmed-operator changeme
(error) WRONGPASS invalid username-password pair or user is disabled.
```

Instead, use your updated password:

```text
$ valkey-cli -h 10.1.44.126 -p 6379
10.1.44.126:6379> AUTH charmed-operator moresecurepassword
OK
```

## Handle multiple passwords

Charmed Valkey maintains multiple internal users with different permissions for
different scopes:

> * `charmed-operator`: the user that manages the database instances

Reviewer: Chore: Empty line around lists.

```suggestion

* `charmed-operator`: the user that manages the database instances
```

Author: done

# How to manage passwords

This guides provides instructions for creating, updating, and otherwise managing passwords.

To read or write data in Valkey, we need to authenticate ourselves.

> 

Reviewer 2: Could you add a section about the self-managed/generated credentials first, so the user knows that we already take care of this if they don't wish so. 
With this, we lower the entry bar to the charm.

Author: I've added a short section for this.

```

Charmed Valkey will now apply the new password to its internal admin user. You can
check the progress by running `juju status`. After a few moments, the deployment will settle:

```text
Model     Controller      Cloud/Region        Version  SLA          Timestamp
tutorial  k8s-controller  microk8s/localhost  3.6.14   unsupported  19:28:26+01:00

App                       Version  Status  Scale  Charm                     Channel   Rev  Address         Exposed  Message
self-signed-certificates           active      1  self-signed-certificates  1/stable  586  10.152.183.111  no       
valkey                             active      3  valkey                    9/edge     11  10.152.183.123  no       

Unit                         Workload  Agent  Address      Ports  Message
self-signed-certificates/0*  active    idle   10.1.44.89          
valkey/0*                    active    idle   10.1.44.126         
valkey/1                     active    idle   10.1.44.117         
valkey/2                     active    idle   10.1.44.127         
```

Now you can use the password to access Valkey. Select the IP address for one of the units to connect to:

> 

Reviewer 2: we do not have nodeport for external access on k8s, so can you add a section for k8s, where you show how to connect using hostnames (you'll ssh into the pod) similar to [here](https://discourse.charmhub.io/t/charmed-mongodb-k8s-tutorial-replica-set-3-access-a-replica-set/15833)

Author: As this is a How-To for managing passwords, I will omit it here and leave it as it is. Instead, I will add this to the section `Access Valkey` in the tutorial, where it is better suited.

# How to manage passwords

> This guides provides instructions for creating, updating, and otherwise managing passwords.

Reviewer: Suggestion: Typo fix.

```suggestion
This guide provides instructions for creating, updating, and otherwise managing passwords.
```

To read or write data in Valkey, we need to authenticate ourselves.

For this guide, we will use Charmed {spellexception}`Valkey's` internal admin user

> `charmed-operator`. This user is only for internal use, and it is created automatically

Reviewer: Question: Can we add a short reference/link to what to do if we need a user for non-internal use? It's OK if we don't have that content yet, but in that case, we should add it later here.

Author: This will be added with the next PR, the feature is currently still in development.

by Charmed Valkey.

We will go through setting a user-defined password for this admin user and configuring
Valkey. 

## Auto-generated credentials

For security purposes, Charmed Valkey automatically generates users and passwords 
for its operations and internal administration tasks. These credentials are stored
in a Juju secret owned by the charm. Inspect the secret:

```shell
juju show-secret valkey-peers.valkey.app.internal_users_secret --reveal
```

```{caution}
This secret is only for internal use. It must not be updated by users.
```

To override the auto-generated passwords for the internal users, follow the instructions
of the next section.

## Configure a user-provided password

First, create a secret in `Juju` containing your password:

```shell

> juju add-secret passwords charmed-operator=changeme

Reviewer: Suggestion: Let's use a proper placeholder here to make it less ambiguous and more visible.

```suggestion
juju add-secret passwords charmed-operator=<NEW_PASSWORD>
```

check the progress by running `juju status`. After a few moments, the deployment will settle:

```text
Model     Controller      Cloud/Region        Version  SLA          Timestamp
tutorial  k8s-controller  microk8s/localhost  3.6.14   unsupported  19:28:26+01:00

App                       Version  Status  Scale  Charm                     Channel   Rev  Address         Exposed  Message
self-signed-certificates           active      1  self-signed-certificates  1/stable  586  10.152.183.111  no       
valkey                             active      3  valkey                    9/edge     11  10.152.183.123  no       

Unit                         Workload  Agent  Address      Ports  Message
self-signed-certificates/0*  active    idle   10.1.44.89          
valkey/0*                    active    idle   10.1.44.126         
valkey/1                     active    idle   10.1.44.117         
valkey/2                     active    idle   10.1.44.127         
```

Now you can use the password to access Valkey. Select the IP address for one of the units to connect to:

```shell
valkey-cli -h 10.1.44.126 -p 6379
```

Authenticate with the username and password you just configured:

```shell

> 10.1.44.126:6379> AUTH charmed-operator changeme

Reviewer: Suggestion: `10.1.44.126:6379>` - Is this a CLI prompt? As a general rule, we should not have them in the code blocks. You can either drop it entirely or use the [terminal block](https://canonical-starter-pack.readthedocs-hosted.com/stable/reference/myst-syntax-reference/#terminal-output) instead if you want to have a CLI prompt here.

Author: I've gone through the docs and used terminal blocks wherever needed.

```

Check the current health of the server with this command:

```shell
10.1.44.126:6379> ping
```

## Update the password

To update your user-configured password, simply update the value of the secret. Here's an example:

```shell

> juju update-secret passwords charmed-operator=moresecurepassword

Reviewer: Suggestion: Use a placeholder with angle brackets here, please. See example above.

```suggestion
juju update-secret passwords charmed-operator=<MORE_SECURE_PASSWORD>
```

```

After running this command, Charmed Valkey will immediately update the password.

> After the deployment has settled again, you can no longer use the old password to

Reviewer: Suggestion: Let's rephrase to avoid repeating "after".

```suggestion
After running this command, Charmed Valkey will immediately update the password.
Once the deployment has settled to `active`/`idle` state again, you can no longer use the old password to
```

access Valkey. Instead, you will receive an error similar to this:

> ```shell

Reviewer: Suggestion: Let's use a text code block to avoid rogue syntax highlighting.

````suggestion
```text
````

(error) WRONGPASS invalid username-password pair or user is disabled.
```

Instead, use your updated password to authenticate:

```shell
10.1.44.126:6379> AUTH charmed-operator moresecurepassword
```

## Handle multiple passwords

Charmed Valkey maintains multiple internal users with different permissions for
different scopes:

* `charmed-operator`: the user that manages the database instances
* `charmed-replication`: the user performs replication between primary and replica instances of Valkey
* `charmed-sentinel-operator`: the user that manages Sentinel for Valkey
* `charmed-sentinel-peers`: the user for communication between Sentinel instances
* `charmed-sentinel-valkey`: the user that Sentinel uses to connect to Valkey
* `charmed-stats`: the user for monitoring and observability

It is possible to manage the passwords for all of above's users with a Juju secret, 
or just for some of them.

To set the password for the `charmed-operator` and `charmed-sentinel-operator` users, 
but keeping the automatically generated passwords for all other users, run the following
command:

```shell

> juju update-secret passwords charmed-operator=moresecurepassword charmed-sentinel-operator=sentinelpassword

Reviewer: Suggestion: Update to use placeholders as above.
